src/lib/steps/package.ts

```diff
@@ -28,105 +28,213 @@ const resolvePaths = (workingDirectory: string, packageConfig: NgPackageConfig):
   }
 
-  const baseDirectory = path.dirname(file);
+  const baseDirectory = path.dirname(filePath);
 
   // read custom ng-package config file
-  let promiseChain = readNgPackageFile(file);
+  let promiseChain: Promise<NgPackageConfig> = readNgPackageFile(filePath);
 
   const defaultPath = path.join(baseDirectory, 'ng-package.json');
-  if (defaultPath !== file) {
+  if (defaultPath !== filePath) {
     // read default ng-package config file
-    promiseChain = promiseChain.then((ngPkg: NgPackageConfig) => {
-
-      return readNgPackageFile(defaultPath)
-        .then((otherNgPkg: NgPackageConfig) => {
-          // merge both ng-package config objects
-          return merge(ngPkg, otherNgPkg, arrayMergeLogic);
-        });
-
+    promiseChain = promiseChain.then(async (ngPkg: NgPackageConfig) => {
+      const otherNgPkg: NgPackageConfig = await readNgPackageFile(defaultPath);
+      // merge both ng-package config objects
+      return merge(ngPkg, otherNgPkg, arrayMergeLogic);
     });
   }
 
-  return promiseChain
-    .then((ngPkg: NgPackageConfig) => {
-      // resolve paths relative to `ng-package.json` file
-      const dir = path.resolve(baseDirectory, ngPkg.src || '.');
+  const ngPkg: NgPackageConfig = await promiseChain;
+  // resolve paths relative to `ng-package.json` file
+  const packageConfigurationDirectory = path.resolve(baseDirectory, ngPkg.src || '.');
+
+  // read 'package.json'
+  log.debug('loading package.json');
+
+  const pkg: any = await readJson(path.resolve(packageConfigurationDirectory, 'package.json'));
+  // merge package.json ng-package config
+  const finalPackageConfig = merge(ngPkg, pkg.ngPackage, arrayMergeLogic);
+  // make sure we provide default values for src and dest
+  finalPackageConfig.src = finalPackageConfig.src || packageConfigurationDirectory;
+  finalPackageConfig.dest = finalPackageConfig.dest || path.resolve(packageConfigurationDirectory,'dist');
+  return new NgPackageData(
+    finalPackageConfig.src,
+    pkg.name,
+    finalPackageConfig.dest,
+    finalPackageConfig.src,
+    finalPackageConfig
+  );
+}
 
-      // read 'package.json'
-      log.debug('loading package.json');
-      return readJson(path.resolve(dir, 'package.json'))
-        .then((pkg: any) => {
-          // merge package.json ng-package config
-          const finalPackageConfig = merge(ngPkg, pkg.ngPackage, arrayMergeLogic);
+async function readSecondaryPackage(rootPackage: NgPackageData, filePath: string): Promise<NgPackageData> {
+  const baseDirectory = path.dirname(filePath);
+  const ngPackageFile = path.resolve(baseDirectory, 'ng-package.json');
+  const packageJsonFile = path.resolve(baseDirectory, 'package.json');
 
-          // read 'ng-package.schema.json'
-          return schemaPromise.then((SchemaClass) => {
-            const schema = new SchemaClass(ngPkg);
+  let ngPackage: NgPackageConfig = await readNgPackageFile(ngPackageFile);
+  const packageJson = await readJson(packageJsonFile);
```

Comment:
We should allow empty `package.json` files. I think some, maybe most, people, will create a file w/o contents and then run in the following error:

 ```
SyntaxError: ...\packages\components\lib\alert\package.json: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
```

(Maybe related to the other error ( `Cannot set property 'externals' of undefined`)

Reply:
Another thinking: `new SchemaClass(ngPkg).$$get('...')` used to do default configuration values. As this got removed. It may have changed default value behaviour!

Reply:
Added a function that returns `{}` instead of throwing when parsing json. I checked the schema and found that only one field was not matching the old default: `lib.entryFile`. I made a change to the root package's parsing so that it defaults to `src/public_api.ts` again, but I'm thinking that the secondary packages should default to `public_api.ts` so that users don't have to change their structure to allow an empty package.json. Since you are the owner, it is ultimately your decision, but I have seen many more secondary packages that do not have a second src folder, and it would make sense to cater to what we think most people will want.

Reply:
Something that just occurred to me: why does `ngPackage.src` default to `.` when `ngPackage.lib.entryFile` defaults to `src/public_api.ts`? Wouldn't it make more sense to have both fields either assume the existence of `/src`  or have both fields assume `.` is inside `/src'?

Rather, it seems odd that the defaults aren't (`.` & `public_api.ts`) or (`./src` & `src/public_api.ts`). Personally, I would prefer the former because that's how most of my libraries are structured.